### PR TITLE
add agent dnsPolicy option

### PR DESCRIPTION
**Pull request by faceair**
ref: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy

It should solve some dns problems, like this https://github.com/jaegertracing/jaeger-operator/issues/676#issuecomment-731601405

### CODE REVIEW

File: pkg/deployment/agent.go [hunk 1]
```diff
@@ -93,8 +93,12 @@ func (a *Agent) Get() *appsv1.DaemonSet {
 	sort.Strings(args)
 
 	hostNetwork := false
+	dnsPolicy := a.jaeger.Spec.Agent.DNSPolicy
 	if a.jaeger.Spec.Agent.HostNetwork != nil {
 		hostNetwork = *a.jaeger.Spec.Agent.HostNetwork
+		if dnsPolicy == "" {
+			dnsPolicy = corev1.DNSClusterFirstWithHostNet
```
Comment (jpkrohling) [inline pkg/deployment/agent.go:100]: Do you need to set a default value here? Is that different than if a value simply isn't provided?

File: Makefile [hunk 2]
```diff
@@ -371,7 +371,7 @@ install-tools:
 		golang.org/x/lint/golint \
 		golang.org/x/tools/cmd/goimports \
 		github.com/securego/gosec/cmd/gosec@v0.0.0-20191008095658-28c1128b7336 \
-		sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.4 \
+		sigs.k8s.io/controller-tools/cmd/controller-gen@v0.3.0 \
```
Comment (jpkrohling) [inline Makefile:374]: This doesn't belong to this PR, I believe.

File: pkg/deployment/agent.go (hunk 1 above)
Comment (faceair) [reply pkg/deployment/agent.go:100]: According to the [k8s documentation](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy),  if we do not set `dnsPolicy` to `DNSClusterFirstWithHostNet`, k8s will use `ClusterFirst` by default, and the daemonset agent will not be able to resolve the service ip of the collector properly.
```
Error while dialing dial tcp: lookup jaeger-collector.observability.svc.cluster.local on {hostIP}:53: no such host
```

File: Makefile (hunk 2 above)
Comment (faceair) [reply Makefile:374]: Yes, it should belong to this [PR](https://github.com/jaegertracing/jaeger-operator/pull/1305).
image
I just don't want to open another pr for this small fix.

File: pkg/deployment/agent.go (hunk 1 above)
Comment (jpkrohling) [reply pkg/deployment/agent.go:100]: Does it mean that DaemonSets aren't working at the moment? I think we have e2e tests exercising this.

File: Makefile (hunk 2 above)
Comment (jpkrohling) [reply Makefile:374]: Alright, that's fine.

File: pkg/deployment/agent.go (hunk 1 above)
Comment (faceair) [reply pkg/deployment/agent.go:100]: Yes, DaemonSets should not be working now. My company has bypassed this problem by forwarding some dns domains to coredns on the physical machine.
I can try adding the corresponding test case to the e2e test.

Comment (jpkrohling) [reply pkg/deployment/agent.go:100]: @kevinearls, we do have this, don't we?

@faceair if you could provide an e2e to prove, that would be awesome!

**Comment by kevinearls** [comment]
@faceair  @jpkrohling  We have daemonset_test.go but I have not seen any failures with it.  Does this mean there could be a problem with the test?

**Comment by obliadp** [comment]
Would love to see this merged. I'm using the daemonset, and agents are not able to find collector unless I patch with `dnsPolicy: ClusterFirstWithHostNet`

File: pkg/deployment/agent.go (hunk 1 above)
Comment (faceair) [reply pkg/deployment/agent.go:100]: I modified the existing es2 test case, is this OK?

Comment (jpkrohling) [reply pkg/deployment/agent.go:100]: ping @kevinearls

**Comment by obliadp** [comment]
Any progress on merging this?

**Comment by jpkrohling** [comment]
ping @kevinearls

**Comment by kevinearls** [comment]
@obliadp  Can you just fix the spelling of "agent-as-sidecar-with-hostnetowk" in test/e2e/sidecar_test.go?

**Comment by obliadp** [comment]
> @obliadp Can you just fix the spelling of "agent-as-sidecar-with-hostnetowk" in test/e2e/sidecar_test.go?

I'm just an interested third party. Your question is probably for @faceair ?

**Comment by D-wenhao** [comment]
@obliadp @faceair @kevinearls @jpkrohling 
How about this problem? The agent can't configure DNS policy. After setting the host network, the agent can't access the cluster DNS. Manual patching can work, but the patch will be removed by the operator after a period of time

**Comment by jpkrohling** [comment]
@faceair are you still interested in this PR?

**Comment by faceair** [comment]
Is there anything else that needs to be changed in this pr? I've fixed the typo.

**Comment by jpkrohling** [comment]
@kevinearls could you please review this one again?

**Review by kevinearls — APPROVED** [review]
LGTM

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
Thanks for your contribution, @faceair!
